### DISCUSSION TITLE
CFP: Enhance Limiting Identity-Relevant Labels user journey

### DISCUSSION THREAD

**Discussion by ovidiutirla** [cilium/cilium#35847] [Ideas]
### Problem: 
Currently, modifying the [cilium/identity-relevant-labels](https://docs.cilium.io/en/stable/operations/performance/scalability/identity-relevant-labels/) filter requires restarting all Cilium agents. As we move towards operator-managed CIDs, this will also necessitate restarting the operator, adding complexity for cluster administrators.

### Proposal: 
Leverage dynamic configuration ([CFP: Dynamic ConfigMap](https://github.com/cilium/cilium/issues/34101)) to allow both agents and operator to watch the labels filter config key and react to changes without requiring a restart. Reconciliation would be triggered by manually modifying the labels filter in the ConfigMap.

### Next Steps: 
I am seeking feedback on this proposal and am happy to draft a more detailed design document if there is sufficient interest. Please share your thoughts and opinions.

### Discussion from Slack so far:

> @aanm: Changing the filter labels for all agents at the same time might have scale implications since they could cause DDoS of the identity-backend. How could we avoid such scenario?
@marseel: I would say if operator manages identities - this is fine in terms of identities. Reconciliation of CiliumEndpoint/CES with new identities is more problematic imho.
@ovidiutirla: I think there are two scenarios here: When Operator Manages CIDs, When Agent Manages CIDs
with 2, we could add a random delay depending on the cluster size; with 1, the identities will be handled by the operator and agents will just wait for the identity propagation, while the old identities will still be present until all agents moves to the new identity. From the above proposal, the labels filter will only change during manual interaction (i.e modifying the ConfigMap). We can also rate limit the number of changes from the ConfigMap to ensure the manual interactions will not overload
@aanm We would need to understand how big the delay is and if it had any security implications associated with it as well as have a way to report that to users.
@ovidiutirla My goal here is to simplify the user journey when updating the labels filter as it becomes more and more complex to update these filters when the CIDs will be managed by Operator. I understand your concern, and I think is totally valid, so I hope we'll get some great ideas to enhance this area. :slightly_smiling_face: Similarly, right now there is manual interaction to restart cilium-agents so the delay will still be there tho, we could expose a metric to highlight the drift delay for applying the propagation at agent side too. Applying the labels update and new identities propagation would still cause less load (from kube-apiserver/KCP perspective and nodes itself as we do not need to re-create all the watchers etc)

> @joestringer: At a high level, this option should not be changing dynamically - we should be able to establish very clear outlines with users when they set up the clusters about which labels are security-relevant or not. I can understand if the wrong decision was made in the first place that having a process to update it would be good, but I don't think that we should be designing the solution to this in order to try to support dynamic updates of this setting on a regular basis.
In the extreme case, it's impossible to reconcile a change in this configuration setting. That's the biggest danger.
My gut feeling would be to tackle this problem more as something we resolve with outside structured tooling to perform the migration rather than integrating it into Cilium to dynamically react to changes in the cluster.
That is, even if the code / process / etc is in the Cilium codebase, build specific tools for the migration rather than expecting cilium-agent/cilium-operator to dynamically react to arbitrary requests to change security-relevant properties.
At some level, cilium-agent would have to learn about the change in security identity for an Endpoint and make that happen, so I guess there is a question about exactly where the line is there and how that could work well
But let's say we have a dedicated tool to assess the change and calculate how the migration is going to work, how the identity pool will be used, and then be responsible for gradually performing the migration.
I think that structure would make it much less likely to shoot yourself in the foot. Another big concern would be if we have to reuse numeric identities, then there needs to be proper separation of the usage to make sure that some agent doesn't make the wrong network policy decision due to the change in meaning for the numeric identity
anyway all of these thoughts are why I think we shouldn't be discussing this on Slack - all of this context will be lost if we don't move it somewhere more permanent.

Let's keep the discussion ongoing and align on the solution.
/cc @aanm @cilium/sig-scalability @joestringer @liyihuang @sypakine @dlapcevic @wedaly

**Comment by liyihuang** [comment]
it's a good idea but I personally think the dynamic configmap will cause more confusing and more problems. 
1. which config can be hot reload and which can't, and it's hard to be remembered. If there are some config can be reloaded and some can't.  I tend to restart the pods anyway to make sure the new config got reloaded.
2. some label change can be very expensive. As far as I know label change on namespace can be very expensive to CPU and memory. Alibaba cloud shared their experience in their kubecon China this year(https://static.sched.com/hosted_files/kccncossaidevchn2024/bf/KubeCon%20China%20%202024%20slides.pdf?_gl=1*1t3rhpw*_gcl_au*OTY4MTY4MzkyLjE3MzA5OTY4OTA.*FPAU*OTY4MTY4MzkyLjE3MzA5OTY4OTA.) search Real world disasters page from the slides. Dynamic config might cause some issues that users might not expect.

Is the concern restarting agent and operator will cause the network disruption?

**Reply by ovidiutirla** [reply]
1. It's not necessarily about remembering, looking up the documentation before changing a config is highly recommended as there could be significant changes anyway in between cilium versions, restarting a cilium-agent for a config that also supports dynamic config ("hot reload') won't hurt.
It's not always feasible to restart all the cilium pods, consider large environments with more than a few thousands of nodes, each time you restart a pod the whole cilium-agent needs to re-init, considering all the informers that need to re-pull and sync, CPU increase at node level etc... and would also want to highlight the downtime until the agent restarts.

2. The same issue you mentioned could happen if you restart all agents and even worse than that. I think this brings even more motivation into finding a solution for this problem rather than just relying on agent restarts. If we can provide a hassle-free solution and take into account this scenarios will offload the stress of changes from the user side. 


-- For this particular scenario with labels filter, I think if we manage to provide a solution that would not rely on user action we could actually reduce the unexpected number of issues. 

Restarting the agent/operator is not always feasible, there is also the need for the user to know how and when to update the labels filter, I also had another discussion around only consider the NetworkPolicy labels for identities which could potentially get rid of labels filter but there was no alignment with the community (https://github.com/cilium/cilium/issues/32576).

**Reply by sypakine** [reply]
> ... I also had another discussion around only consider the NetworkPolicy labels for identities which could potentially get rid of labels filter but there was no alignment with the community (#32576).

While conceptually the cleanest approach, I understand the hesitation with this approach. It shifts the responsibility from the (cluster) admin persona to application operators (namespaced) that cannot be expected to know the indirect consequences of network policy label selectors. This would likely result in noisy-neighbor problems.

**Reply by liyihuang** [reply]
Sorry, I got some issues with my GitHub notification. I personally would prefer the assumption that the cilium agent is managed by the cluster admin, we should prevent the app or namespace admin creating a lot of noises. Even restarting the agent is not ideal but it's better than random changing labels causing the possible disaster for the cluster.

**Comment by joestringer** [comment]
> At a high level, this option should not be changing dynamically - we should be able to establish very clear outlines with users when they set up the clusters about which labels are security-relevant or not. I can understand if the wrong decision was made in the first place that having a process to update it would be good, but I don't think that we should be designing the solution to this in order to try to support dynamic updates of this setting on a regular basis.

I wanted to pick this thread up a bit more to expand on it. In general, a Security Identity should at its core highlight a specific security domain (set of workloads) that is different from other security domains, so we can then enforce network policies between those security domains. Based on the architecture of your deployed workloads, the underlying relationship between these workloads is not shifting all the time, so similarly the Security Identity should not be shifting all the time. Perhaps you have a microservice architecture and you introduce a new workload as part of the architecture - that's one new Security Identity with a specific relationship to other existing Security Identities in the environment. This is not shifting the way that existing Security Identities are defined.

When we talk about configuring the identity-relevant labels filter, this is defining which labels in a k8s environment are used to define the security domains of the applications in the environment. In an ideal world I think we should have this very clearly defined with users up-front. There's already recommendations in the doc about the sorts of things that do not make sense for defining the security domains: timestamps, hashes, UUIDs, that sort of thing. 

I can understand that there are nuances to this and for many newer users, they may not be aware when they first deploy Cilium. So from that perspective, I can understand if later on the user realizes that the labels filter is causing Cilium to operate suboptimally based on the label usage for workloads in their environment. Having a way to safely transition at that point makes sense to me. But I see this as being in the category of "I do this maybe once or twice when I learn more about Cilium or my organization makes a significant shift in the way it uses labels", and not "we need this to be highly dynamic to change the meaning of Security Identities in my environment multiple times per month or more frequently".

**Reply by sypakine** [reply]
The term dynamic conjures the notion that mutations will be frequent. I  prefer to think of the proposal as making this field declarative -- admins need-not be concerned with how the field gets reconciled.

I agreed that this field should not be mutated frequently, and concede that making the process streamlined for administrators increases the probability of misuse (as restarting workloads is a signal that gives users/admins pause).

I would be interested to know what other platform/infrastructure operators (AKS, EKS, etc.) think of the proposal. For GKE, the tooling instrumentation discussed in chat would be challenging, but we are also cognizant that should only be context and not a strong motivator for pursuing this change.

**Comment by joestringer** [comment]
I would also like to understand a bit more about if there are common usage of labels that cause problems with the labels filter, or if there are integrations with specific software where that software creates labels that should not be considered for Security Identities. If we can solve part of this problem with better defaults in the cilium-agent, I think that can mitigate some of the need to be able to dynamically modify the filter.

If anyone has examples of labels that Cilium considers security-relevant by default which Cilium perhaps shouldn't take into account for Security Identities, please share in this thread.

**Reply by wedaly** [reply]
We had one feature request in AKS to configure this, because the customer installed trivy-operator which uses unique labels for each job it creates: https://github.com/Azure/AKS/issues/4608

**Reply by sypakine** [reply]
In GKE we have observed this with Istio `service.istio.io` labels and Jenkins labels; `service.istio.io/canonical-name`, `service.istio.io/canonical-revision`, and `jenkins/label-digest` being the most problematic.

I think addressing with defaults makes sense for k8s APIs but won't scale well for the wider OSS ecosystem.

**Reply by joestringer** [reply]
> I think addressing with defaults makes sense for k8s APIs but won't scale well for the wider OSS ecosystem.

If we can catch 90% of cases, I think that's already huge. I also think this scales reasonably well once we've got into the habit of updating this default; it should be maybe a 5-10 line patch with the extra labels and a docs update. We're probably talking about dozens not hundreds of very common software components.

**Reply by ovidiutirla** [reply]
I tend to agree with @sypakine that manually adding third parties to the defaults won't fix the problem in the long run. Firstly, there are just too many third parties to cover. The ones mentioned are just two of them, and we already have domain knowledge about them. Consider the case of a third party where we must invest time to understand how it works and which labels should be properly excluded. Excluding third-party labels in our defaults doesn't sound like the right approach. 

Also, think from the user POV, on Cilium 1.17 there are no defaults for this specific third party and they create their workloads based on this assumption, on 1.18 we magically changed that, I think we are talking about more than hundreds tho... We force the users to spend more time reading the documentation and understanding complex scenarios rather than optimizing their journey and providing hassle-free solutions.

**Reply by wedaly** [reply]
Another example we saw recently in AKS: spark-operator puts UUIDs in labels, which caused an outage for one of our customers.

**Reply by wedaly** [reply]
it might be a wider discussion, but I'm wondering if with operator-managing identities enabled Cilium might be able to detect and report on identity labels with high cardinality, so customers at least have some indication that this is happening.

**Comment by joestringer** [comment]
When solving this problem, what do we think the ideal architecture/design looks like?

I think that these are our base assumptions which we can use to explore the intended design:
1. There is a set of Pods that determine the full label set of each workload in the cluster.
2. There is a labelsfilter which determines which labels are important for security.
3. There is an "Identity Manager" logical component that goes through every unique Pod labels set, applies the filter, and the result is the Security Identity.
4. Each unique Security Identity (label set) is assigned a unique number, synced centrally. This helps to make network policy enforcement more efficient by having a machine-readable short-hand for the same concept.
5. Every node must ensure that the policy enforcement for a Pod uses the corresponding Security Identity, that is, network policies applying to the Pod based on the labelsfilter will apply.
6. Every node must also ensure that when a local Pod connects with a remote Pod, connections to the remote peer are also enforced with the corresponding Security Identity for the peer (and the same for connections in the opposite direction).

Currently we implement (3) as a distributed algorithm by giving that responsibility to cilium-agent on each node, though ultimately to complete the step it requires centralized synchronization, in part due to step (4). If the number of unique Pods / unique labels is very high in (1), and (2) changes which labels matter, then that means that (3) can generate a lot of work. Not only that, but each individual instance does not know the scope of the overall problem, so they may all work independently to attempt to transition every identity to a new {labelset,number}. Some of the work is duplicated as the same workload may be present on multiple nodes, but also depending on how we model the transition, in the worst case the re-allocation of security identities could allocate new identities for every single workload, significantly increasing the demands on the security identity space[^1]. Furthermore, it's hard to understand how this impacts step (6) which involves plumbing policymaps for each Endpoint based on the network policies applied in the cluster and the security identities currently allocated.

> When Agent Manages CIDs [...], we could add a random delay depending on the cluster size

Cluster-dependent ratelimiting of the process overall can mitigate some of the impacts of this change, but I don't think that this provides good bounding of worst-case behavior. I think it's worth discussing whether we want to solve this problem with the current distributed identity allocation algorithm or not. This can make a big difference to how we model the overall transition from one labelsfilter to another.

I think it's a fair assumption to start with a cluster where the number of security identities is not causing problems. I think we're assuming that the target state after labelsfilter reconfiguration would also be a likewise stable state, though we actually don't know this. In the worst case, a user inadvertently may include labels that cause high Security Identity cardinality, leading to bad things :tm:. I would feel a lot more confident if the migration process could perform some sanity checks like this rather than just assuming that the start and end state are both valid and that the transition can just be applied. Then beyond this, there is the intermediate state which is likely to increase pressure on the numeric identity space and perhaps the policymap content for individual Cilium Endpoints on different nodes. This could be as minimal as "no effect whatsoever" or as invasive as "this will cause connectivity outages for endpoints with permissive policies" (hopefully only temporary).

[^1]: Worst case if all nodes can't properly coordinate, each node races to allocate their own security identities for each unique label set present on its own node---a result of worse than doubling the number of security identities.
